### ISSUE TITLE
[3.2.1 regression] Shortcut .lnk files cannot be synced anymore

### ISSUE THREAD

**Issue by biva** [nextcloud/desktop#3244]
I just installed 3.2.1 and I have the feeling that previously, .lnk files (shortcut in Windows) were synced (at least that I could remove these files from the **Ignored Files** list.

On 3.2.1, I can't force the sync anymore.

Am I wrong?

**Comment by biva** [comment]
Actually, I confirmed it: previous versions allowed to sync .lst files. Is it possible to solve that?

**Comment by FlexW** [comment]
Yes, it was possible to sync these files in the past, but they caused trouble with the virtual files. Since the Nextcloud desktop client is intended to be used cross platform, we also try to not sync any platform dependent files, because they have no meaning on other platforms. For example it is not possible to use/create .lnk files on macOS or Linux.

Please have a look at the follwing pull request: https://github.com/nextcloud/desktop/pull/3057

**Comment by biva** [comment]
@FlexW Thanks for the explanation, I understand that virtual files create an additional issue. As @er-vin mentioned in a comment, as a user using .lnk files (only between Windows systems), I would have appreciated to see a warning before installing it (in the changelog for instance).

Also, is there a possibility to see this solved in the future? (to be able to sync last files again) In this case, shouldn't we leave this issue open?

**Comment by FlexW** [comment]
@biva Yes a warning would have been better. This can also cause issues when using Windows only. The problem are lnk files that point outside of the sync folder.

**Comment by biva** [comment]
Thanks for your kind answer @FlexW 
I read in your post that _.lnk files will be treated as regular files_ (in a future version of Qt). Does it mean that we'll be able to sync .lnk files again? If yes, any idea of the timeline?

**Comment by FlexW** [comment]
@biva No, we don't plan to sync the .lnk files again. In my opinion it was a bug to even allow to sync these files in the past. We don't allow to sync UNIX's symbolic links as well for the same reasons. If we allow .lnk files, then we also should allow to sync symbolic links. If we want to do that in a proper way, we would need to make sure that .lnk files get replaced on UNIX systems with symbolic links and the other way round. To do this right, it would require a lot of work on the sync engine (and maybe the server). For now we want to focus on improving and stabilising the current sync engine and not introduce some new fancy feature that probably not enough people use to make it worth investing the time :) This might change if in the future a lot of people (especially paying customers) require that feature.

**Comment by biva** [comment]
What about for users who sync those files only on Windows? I really think it's a pity to not allow users to choose what they want to sync or not. I think that a good way would be to not sync these files by default to avoid issues, warn people about the consequences, and let users choose if they want to sync them or not.

We are syncing a lot of lnk files without any issues on 20 computers: your decision to remove this possibility puts us in a bad situation and we don't know how we'll reorganise all our processes.

If the issue you're mentioning is the same on 3.2 than on 3.1, and that having this possibility was just a mistake, then we would like to have this mistake again in 3.2, as an option. Is it doable?

**Comment by FlexW** [comment]
The problem is also on Windows. The problem are .lnk files that point outside of the sync folder. Maybe we could discuss putting .lnk files into to the sync exclude list and then you could remove them from the exclude list and sync them. @allexzander @mgallien @camilasanI know we said it would be better to hardcode it in the code but what do you think?

**Comment by allexzander** [comment]
@FlexW @biva We can discuss the possibility to bring .lnk syncing back, but, only when the VFS is disabled. Otherwise, showing a warning for .lnk files with VFS turned on is simply not enough. The desktop client is simply freezing and is completely unusable when having the .lnk files enabled in VFS mode. @biva you can observe this issue if using the latest 3.2.0 RC 3 from here https://download.nextcloud.com/desktop/prereleases/Windows/. Severity-wise, it is way worse than simply ignoring .lnk files.

But yeah, we may consider allowing this for non-VFS mode. But, this may also bring a problem of inconsistent sync engine behaviour between VFS and non-VFS modes.

**Comment by biva** [comment]
Hello @FlexW and @allexzander 

Thank you very much for your effort to find a solution. I understand the technical issue behind your decision.

I think allowing this for non-VFS mode is already a good start: users know the risk and they will have to make manual action to allow it. I think nobody can complain if it leads to issues.

For VFS-mode, if I understand the future version of Qt's statement ("_.lnk files will be treated as regular files_"), maybe you can allow it for windows users, when it will be possible with Qt?

For us, the VFS-mode is more interesting than the lnk management: we are in Africa and the bandwidth is expensive and slow. This is really useful for us to use VFS-mode. We'll wait for a solution for lnk with VFS.

**Comment by haniham** [comment]
I hope there will come a solution - so far i can't fix it by myself

I am using nx-cloud to backup files on my parents desktop so files with a big yellow exclamation point raise questions which turn to my problems for me. For me the easiest way would be that the link files keep on being synced as any other files, with the possibility for the ignore list,... . The link-follow is nice but unneeded for me. The problem is that the files are ignored without me being able to control them being ingnored.

And the rollback to older versions hasn't worked out quite well, so im in real need for a fix in future versions.

**Comment by ChrisBiesterveld** [comment]
Hi,

Landed on this page to find out why .lnk files where not synchronizing anymore and looked at the answers provided.

I do not understand the technical issue explained allexzander as it out of my expertise. Although the reasons could be very valid, from a user perspective hosting my own environment, I would like to have control over what I want to sync or not.

In general I would also appreciate to have the option to just use an exclusion list and not hardcode extensions to give that control back to the user (maybe only on server/admin level ?)

For now it would be sufficient for me if the .lnk files could be included again in the exclusion list. 

Thanks :)

**Comment by Temtaime** [comment]
I think the issue should not be closed until the regression is fixed.
I don't see any reasons why lnk files cannot be synced even with vfs.
.lnk is an ordinary file, it is just treated differently by Windows, so there's no problem with having .lnk files in macos/linux.

**Comment by allexzander** [comment]
@Temtaime It has been explained already why this can not be done. There is a comment in the issue you've just reopened https://github.com/nextcloud/desktop/issues/3244#issuecomment-830190844.

**Comment by haniham** [comment]
But i want to sync the files as files and not follow them.

Where do we come when owncloud decides to exclude more file extensions they don't like from owncloud. That user paternalism is horrid.

**Comment by FlexW** [comment]
@haniham Please understand that the Windows system automatically follows the lnk files when using the virtual file system on Windows. Thats the problem and the desktop client can not prevent that Windows follows these lnk files when implicitly hydrating (downloading) files.

**Comment by ChrisBiesterveld** [comment]
> But i want to sync the files as files and not follow them. It's a file like any other and thus is syncable.
> 
> Where do we come when owncloud decides to exclude more file extensions they don't like from sync in owncloud. That user paternalism is horrid.

I can copy lnk files from one system to another. Without any issues, even if the .lnk file points to something that does not exist. So I do get your response haniham.

I do not understand the "issue" described by FlexW, but that is mainly because I do not understand the technical issue described. If it causes problems for others then so be it, but I did request to have an option to determine it myself by putting it back in the exclusion list or maybe a new option on server level. For me Syncing .lnk files has never been a problem. Manually or via NextCloud. 

So I do hope something will be created to get that functionality back and place the decision on what to sync back to the owner.

**Comment by Temtaime** [comment]
@allexzander  i don't think that windows does anything regarding lnk files.
Linked PR in the comments states that there's QT api which follows lnk files on windows and it is deprecated. There must be other ways getting correct information.

**Comment by zlydk** [comment]
Please consider bringing .lnk syncing back.
My usecase: I sync (many) .lnk shortcuts to network resources across (windows) workstations using NC.
Maybe put .lnk files into to the sync exclude list by default, and let users make an active choice to remove it - with a warning or whatnot.

**Comment by vgdh** [comment]
I want syncing LNK files too.
It would be really cool!

**Comment by wlnx** [comment]
> @haniham Please understand that the Windows system automatically follows the lnk files when using the virtual file system on Windows. Thats the problem and the desktop client can not prevent that Windows follows these lnk files when implicitly hydrating (downloading) files.

Maybe I miss something, but OneDrive client syncs the files with no issue.
Furthermore, I cannot cd into, e. g., Windows.lnk => *.lnk files are not like symlinks that are not regular files at all.
Anyway, I also have file sets with lnk files. And if nextcloud knows what I want to sync better than me... Sad but true.

**Comment by allexzander** [comment]
@wlnx OneDrive is not using Qt. The problem is not in Windows .lnk files or Nextcloud, it's a problem with how Qt version prior to 6.x handles .lnk files on Windows. It is explained earlier in this thread.

**Comment by zlydk** [comment]
> @wlnx OneDrive is not using Qt. The problem is not in Windows .lnk files or Nextcloud, it's a problem with how the Qt versions prior to 6.x handles .lnk files on Windows. It is explained earlier in this thread.

I think everyone agrees, that this problem originates in QT.
The question becomes whether Nextcloud is able (or willing) to look into a possible workaround for the .lnk handling issue in QT.
Some ideas was raised in https://github.com/nextcloud/desktop/pull/3057

I think this issue should be re-opened, as it is valid as reported.

**Comment by wlnx** [comment]
> @wlnx OneDrive is not using Qt. The problem is not in Windows .lnk files or Nextcloud, it's a problem with how the Qt versions prior to 6.x handles .lnk files on Windows. It is explained earlier in this thread.

No doubt this issue is from Qt. I just try to say this is bug, not feature.

**Comment by vgdh** [comment]
And it is a major bug.

**Comment by vgdh** [comment]
It is not expected behavior from file synchronization software.

**Comment by sagitariozod** [comment]
I also want to sync LNK files.
It would be great!
I understand that an .lnk file contains data, it is not the same as a symbolic link, that it is dependent on the operating system, and therefore it should just sync and I should be able to choose whether or not it syncs.
If the problem is QT, perhaps an alternative would have to be considered, although I suppose it would take a lot of effort, but if its use involves restrictions that prevent the user's total freedom to synchronize any type of file that contains data, this should be considered very seriously. change.
My opinion is that something is being done wrong in the desktop client, and I have many years of experience as a computer technician, in fact currently in linux I use the owncloud client because it works better than nextcloud, my server of course has nextcloud installed.
On the other hand, the size of the desktop client installation has become too large, 200mb, for the functionality it has to offer, in my opinion. I just installed google drive, which I haven't used for a long time, install size 60mb and it syncs lnk files perfectly.

Sorry, my English is very bad.

**Comment by ACiDGRiM** [comment]
This is a useful feature.
At my employer we use OneDrive and the users' desktop, including their .lnk files, are sync'd to their cloud storage.
The benefit to users is that when the move to a new computer and login to a fresh profile, all they have to do is sign into the sync client (in this case Nextcloud Desktop) and their desktop shortcuts are available to them.

At least in OneDrive, we do not have issues with .lnk files syncing, even if they point to a local application, a non-existent path, or even a network resource. The concern about .lnk files syncing to Linux or Mac clients could be mitigated by setting those files in the Ignored list on those platforms, and excluding it on windows platforms.

Furthermore, windows users can make the same kind of FS shortcuts linux can in the form of Junctions (limited to non-admin users) which are not sync'd for obvious reasons. This should have no bearing on .lnk sync on windows clients.

**Comment by jdunn0** [comment]
@FlexW said
> No, we don't plan to sync the .lnk files again and that's not bound to Qt. In my opinion it was a bug to even allow to sync these files in the past. We don't allow to sync UNIX's symbolic links as well for the same reasons. If we allow .lnk files, then we also should allow to sync symbolic links.

You are incorrect about .lnk files being the same symlinks.

As someone else mentioned .lnk files are the same as .desktop files on *nix systems. They are regular files which the Windows Shell or a Desktop Environment understands and does something with like launching an app or open a folder specified in the file.
Symlinks as entries in a file system that point to a file to folder which are transparently followed by any application on the system.

Windows did not support symlinks before Windows Vista and that is probably why things like Qt used .lnk in place of them but that is not their purpose.

The files .lnk, .desktop along with others like .url serve a different purpose then symlinks and should be treated as the regular files they are and synced like any other file.

**Comment by biva** [comment]
@jdunn0 thanks for your input. I think that the Nextcloud team is doing its best to solve this issue, and I understand that it is somehow planned for this year, as @allexzander shared in https://github.com/nextcloud/desktop/pull/3057#issuecomment-885433569
> _At some point (could be even this year), we plan to approach this problem from another perspective and bring back .lnk files synchronization._

As far as I understood, you can sync lnk if you don't use vfs. If you want to use vfs and lnk together, like me, we'll have to wait.

I think the Nextcloud team initially didn't see the significant number of valid use cases that a lot of people have with lnk files. These examples have been shared several times here, or in https://github.com/nextcloud/desktop/issues/3499 and in https://github.com/nextcloud/desktop/pull/3057 . I think that it has been accepted by the Nextcloud team as a regression that should be fixed.

To summarise, I think the current situation is:
1. lnk and vfs are not compatible for now, because of technical issues
2. Without vfs, you can sync lnk
2. Users have proven than lnk files should be considered as normal files and be "syncable"
3. The Nextcloud team has a lot of priorities and work, and is doing its best to solve this issue this year

Am I correct? In any case, I think this issue should remain open somewhere, either reopen this one, or merge with https://github.com/nextcloud/desktop/pull/3057 to avoid duplicates. @allexzander and @FlexW would you consider that?

**Comment by CombeeMike** [comment]
@biva I don't think that there's a way for the user to "opt out" of `vfs` and "re-activate" `*.lnk` synchronisation but I might be wrong 🤷‍♂️.

Meaning: Everything you said sounds true except the fact that users have no way ATM to sync `*.lnk` files anymore until the Nextcloud team (hopefully) fixes this somehow.

If I'm wrong: Could you explain how I could achieve this (where to deactivate `vfs` etc.)?

**Comment by wlnx** [comment]
> @haniham Please understand that the Windows system automatically follows the lnk files when using the virtual file system on Windows. Thats the problem and the desktop client can not prevent that Windows follows these lnk files when implicitly hydrating (downloading) files.

No it doesn't. Look and sorry my russian:

Windows: I can cat the lnk-file to terminal, even if it is bound to directory

```
z:\tmp\RR>type test.lnk
L☺¶☻└F(cOБОж╫☺╝гчВОж╫☺(cOБОж╫☺☺☺¶▼PрO╨ ъ:i►в+00Э↓/Z:\J1∟S╜Шtmp8 ♦я╛ШJ;д*Soо.o!☺V☺tmp↕H1╕RIдRR6  ♦я╛╕RНб*S=Ю.ФФ
ПаБRR↕N1*Sсоtest:       ♦я╛*Sсо*Sсо.х7☺
z:\tmp\RR>type test
Отказано в доступе.
```

Linux: I can cat neither directory nor symlink to it, because symlink is a special file system object:

```
~$ ls -ald test*
drwxr-xr-x 2 lynx lynx 4096 Sep 11 01:00 test
lrwxrwxrwx 1 lynx lynx    4 Sep 11 01:00 test.lnk -> test

~$ cat test.lnk
cat: test.lnk: Is a directory

~$ cat test
cat: test: Is a directory
```

Thus, lnk-files are just blobs that is handled in some way by Windows shell "explorer.exe", even cmd or powershell have nothing to do with them.

**Comment by wlnx** [comment]
> Windows did not support symlinks before Windows Vista

Being honest it did. NTFS supported symlinks from Win2000, but symlinking was hidden. In Vista we met mklink utility not to dive into fsutil or something more interesting.

**Comment by wlnx** [comment]
> @biva I don't think that there's a way for the user to "opt out" of `vfs` and "re-activate" `*.lnk` synchronisation but I might be wrong 🤷‍♂️.
> 
> From [here](https://github.com/nextcloud/desktop/issues/3244#issuecomment-916743472):
> 
> > ...
> > As far as I understood, you can sync lnk if you don't use vfs. If you want to use vfs and lnk together, like me, we'll have to wait.
> > ...
> > ```
> > 2. Without vfs, you can sync lnk
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ...
> 
> IMHO everything you said sounds true except the 2 statements above.
> 
> If I'm wrong: Could you explain how I could achieve this (where to deactivate `vfs` etc.)?

I can explain. Just use NextCloud client 3.1.3. It disables vfs and enables lnk.

**Comment by jdunn0** [comment]
@wlnx said
> Being honest it did. NTFS supported symlinks from Win2000, but symlinking was hidden. In Vista we met mklink utility not to dive into fsutil or something more interesting.

Windows 2000 and Windows XP had support for NTFS Junctions, not symlinks.
Junctions could be linked to another folder on the same drive or a different drive with an absolute path but with symlinks you could do relative links to folders, links to UNC paths and links to files.

**Comment by wlnx** [comment]
> @wlnx said
> 
> > Being honest it did. NTFS supported symlinks from Win2000, but symlinking was hidden. In Vista we met mklink utility not to dive into fsutil or something more interesting.
> 
> Windows 2000 and Windows XP had support for NTFS Junctions, not symlinks.
> Junctions could be linked to another folder on the same drive or a different drive with an absolute path but with symlinks you could do relative links to folders, links to UNC paths and links to files.

I ain't ready to setup XP back to check if I'm missing something, thus let it be. It's been a long time ago. =^>.<^=

**Comment by jdunn0** [comment]
> I ain't ready to setup XP back to check if I'm missing something, thus let it be. It's been a long time ago. =^>.<^=
You don't need to setup Windows XP to check as on [Microsoft's official documentation on Symlinks](https://docs.microsoft.com/en-us/windows/win32/fileio/symbolic-links) it says "Symbolic links are available in NTFS starting with Windows Vista".

**Comment by wlnx** [comment]
> > I ain't ready to setup XP back to check if I'm missing something, thus let it be. It's been a long time ago. =^>.<^=
> 
> You don't need to setup Windows XP to check as on [Microsoft's official documentation on Symlinks](https://docs.microsoft.com/en-us/windows/win32/fileio/symbolic-links) it says "Symbolic links are available in NTFS starting with Windows Vista".

Well, let's begin [docufighting](https://en.wikipedia.org/wiki/NTFS_links):
From NTFS 3.1 onwards, symbolic links can be created for any kind of file system object. NTFS 3.1 was introduced together with Windows XP
...
Since Windows XP uses the same NTFS format version as later releases, it's feasible to enable symbolic links support in it. For using NTFS symbolic links under Windows 2000 and XP, a third-party driver exists that does it by installing itself as a file system filter.

I don't think the holywar makes any sense here. Anyway, symlinks and lnk-files are not the same. We all say hello to QT.

**Comment by jdunn0** [comment]
@wlnx It's not a holywar, I was just pointing that there was no built-in support for Symlinks in Windows 2000 and Windows XP. The third-party driver didn't even exist until many years after the release of Windows Vista.
Windows 2000 and Windows XP did have support for Junctions but to create them you needed a third-party tool like Sysinternals Junction.

The entire point of my first point in this topic was just that symlinks and .lnks are not the same.
Qt has been around for a very long time, it may have done the treat .lnk files as symlinks before Windows 2000 even existed so there was likely no Junctions or Symlinks available when that was done.

**Comment by wlnx** [comment]
> @wlnx It's not a holywar, I was just pointing that there was no built-in support for Symlinks in Windows 2000 and Windows XP. The third-party driver didn't even exist until many years after the release of Windows Vista.
> Windows 2000 and Windows XP did have support for Junctions but to create them you needed a third-party tool like Sysinternals Junction.
> 
> The entire point of my first point in this topic was just that symlinks and .lnks are not the same.
> Qt has been around for a very long time, it may have done the treat .lnk files as symlinks before Windows 2000 even existed so there was likely no Junctions or Symlinks available when that was done.

What is "built-in"? FS had full support of symlinks starting with XP (2000 had only junction points, yes). Usermode - yeah, had none. We finished at the start point: before Vista symlinks were available - but only with 3-rd party drivers. Vista brought us usermode and mklink. And Win10 made non-privileged users able to use the functionality. And Qt does strange things.

And no matter what point Qt uses to treat lnk-files as symlinks. The files are not and have never been.

**Comment by jdunn0** [comment]
> What is "built-in"? FS had full support of symlinks starting with XP (2000 had only junction points, yes). Usermode - yeah, had none. We finished at the start point: before Vista symlinks were available - but only with 3-rd party drivers. Vista brought us usermode and mklink. And Win10 made non-privileged users able to use the functionality. And Qt does strange things.

Despite the File System having the capability of symlinks, because Windows XP had no included driver for handling them or APIs for creating them, it wasn't a built-in feature of the OS.
I would say that built-in would refer to things the OS can not do on it's own without third-party drivers.

Back on topic, I tried downloading the OwnCloud Desktop Client and found that it does support syncing .lnk files with Virtual File Support enabled which was a bit of a surprise as I would have thought that both clients would have the same issue as they both have a lot of code in common.

**Comment by wlnx** [comment]
> > What is "built-in"? FS had full support of symlinks starting with XP (2000 had only junction points, yes). Usermode - yeah, had none. We finished at the start point: before Vista symlinks were available - but only with 3-rd party drivers. Vista brought us usermode and mklink. And Win10 made non-privileged users able to use the functionality. And Qt does strange things.
> 
> Despite the File System having the capability of symlinks, because Windows XP had no included driver for handling them or APIs for creating them, it wasn't a built-in feature of the OS.
> I would say that built-in would refer to things the OS can not do on it's own without third-party drivers.
> I can see how this can be interpreted differently though so I guess agree to disagree here.
> 
> Back on topic, I tried downloading the OwnCloud Desktop Client and found that it does support syncing .lnk files with Virtual File Support enabled which was a bit of a surprise as I would have thought that both clients would have the same issue as they both have a lot of code in common.

Ah, I misunderstood you, sorry. Ok, I agree, despite XP had the capability, interaction with it  was a very strange magic, thus being unavaiilable for the most of users.

OwnCloud doesn't meet the issue? xD Maybe it has sense to move from NextCloud.

**Comment by Temtaime** [comment]
@wlnx hello. I'm currently working on custom desktop client that will for sure sync .lnk files. And be faster than current one. :) Stay tuned

**Comment by ScarUY** [comment]
> @wlnx hello. I'm currently working on custom desktop client that will for sure sync .lnk files. And be faster than current one. :) Stay tuned

I'd like to know more! I do need those .lnk files! (currently I'm trying to keep my company's workflow with batch scripts, it's making me hate my life) My team is starting to regret nextcloud and want to go back to owncloud, please help me disuade them, bring back the lnk!

**Comment by biva** [comment]
Hello, if this issue is creating problems in an atomic factory ☢️, then we have big troubles 😅.

Thanks @AtomicWerks for your report. To avoid misunderstanding, this ticket is still open. I guess nobody from the community nor from Nextcloud had the time/capability to solve it yet.

@FlexW I understand that you're considering a "special utility function for this case". I agree that it would be an easier way to go, even if less clean. Just to have an idea, would it be possible to share if Nextcloud moved to QT6.2? And if you're still planning to solve this issue at some point? 

Thanks!

**Comment by jochenwezel** [comment]
### Feature request 1

> But yeah, we may consider allowing this for non-VFS mode. But, this may also bring a problem of inconsistent sync engine behavior between VFS and non-VFS modes.

It would be great to get this feature soon! 
We also share .lnk files and want to to continue using it. 

We don't need that feature of virtual files, but when we need it in future, it's okay to not sync .lnk files. 

### Feature request 2

I recommend a warning info "Please note: syncing of shortcut (*.lnk) files not supported" with a tooltip text "Virtual files feature doesn't support syncing of *.lnk files" at the position marked in yellow in following screenshot:
![image](https://user-images.githubusercontent.com/3033827/152385878-c099336a-d7e1-474c-954f-0bd5c37de514.png)

For Linux/Mac platforms, the warning text should be adapted to match to the platform.

**Comment by ScarUY** [comment]
> ### Feature request 1
> > But yeah, we may consider allowing this for non-VFS mode. But, this may also bring a problem of inconsistent sync engine behavior between VFS and non-VFS modes.
> 
> It would be great to get this feature soon! We also share .lnk files and want to to continue using it.
> 
> We don't need that feature of virtual files, but when we need it in future, it's okay to not sync .lnk files.
> 
> ### Feature request 2
> I recommend a warning info "Please note: syncing of shortcut (*.lnk) files not supported" with a tooltip text "Virtual files feature doesn't support syncing of *.lnk files" at the position marked in yellow in following screenshot: ![image](https://user-images.githubusercontent.com/3033827/152385878-c099336a-d7e1-474c-954f-0bd5c37de514.png)
> 
> For Linux/Mac platforms, the warning text should be adapted to match to the platform.

Where do I vote for this?

**Comment by chylex** [comment]
I have reverted #3057 in the 3.3.6 branch and pushed it to my fork as [stable-3.3-lnk](https://github.com/chylex/Nextcloud-Desktop/tree/stable-3.3-lnk), now my `.lnk` files are syncing again.

I do not use the virtual files feature (it's not even appearing in my client, maybe it's not supported on Windows 7) so I don't care about issues that reverting #3057 will cause with VFS, I just wanted to get `.lnk` files working without downgrading to very old versions.

While I would normally recommend validating my changes and building the branch yourselves, it took me several hours to get a working build due to outdated build instructions and many unhelpful errors, so I have [published the built files](https://github.com/chylex/Nextcloud-Desktop/releases) in case anyone *really* wants to sync `.lnk` files and doesn't want to deal with building. If you decide to use it, make sure you read the caveats, and do not bother the Nextcloud team with any issues that happen while using my build.

**Comment by ChrisBiesterveld** [comment]
Although I appreciate the effort from chylex, it's not usable from a support and maintain perspective. The caveats also instructs the user to create regular backups, one of the main reasons for me to use Nextcloud.

It does show however that the change can be reverted, and that's a great prospect. I personally will just wait until the newest build supports it again. (Hopefully very soon). Until that time I just created a small script to make a frequent copy manually.

**Comment by chylex** [comment]
I do not consider Nextcloud to be sufficient backup. If data corruption happens on either the client side or server side, it will be automatically propagated everywhere with no recovery. If you care about your data, either your server or client (or both) should have off-site backups regardless of whether you use my custom build or not.

**Comment by ChrisBiesterveld** [comment]
You are absolutely correct chylex. It's not what I meant to say. I use Nextcloud as sync and nothing else. For the backup's I was referring to the manual backup of .lnk files. I thought the caveat description of manually creating backup's was related to that.

Bottom line : I misinterpreted your explanation :-)

**Comment by chylex** [comment]
I adjusted the wording, hopefully it's more clear now.

**Comment by mgallien** [comment]
to the best of my knowledge Qt did not change behavior with respect to `.lnk` files.
they can till trigger implicit hydration of the file they point to (if it is also in the sync folder and still pure virtual).
our current desktop client cannot yet deal with itself triggering implicit hydration
we would need to either ensure our code can never access such `.lnk` files if they point to a pure virtual file or build a new architecture that would allow the desktop client to trigger implicit hydration without problems

**Comment by mgallien** [comment]
that said, `.lnk` files should better stay out of a sync folder because there is no way to deal with them in the web UI or on other platforms (Androit mobile client, ...). They are not going to be really usable out of a desktop sync folder on Windows.

**Comment by zlydk** [comment]
> that said, `.lnk` files should better stay out of a sync folder because there is no way to deal with them in the web UI or on other platforms (Androit mobile client, ...). They are not going to be really usable out of a desktop sync folder on Windows.

Using that argument, you might as well exclude syncing .exe and .dll - they are not going to be really usable outside of a desktop sync folder on windows......

I think there are plenty of valid use cases for syncing .lnk that has been described in this thread though?

**Comment by s3woz** [comment]
Personally, I have both many Linux and Windows machines. I'm daily using synced .lnk files not only to link to common folders, but also to URLs of various web apps and notes in Office online. Because of this regression issue, I'm still using nextcloud 3.1.3 client on Windows. Personally, I wish that Linux file managers would also include .lnk support, but this is a completely another story. Anyway, syncing .lnk files with nextcloud is a very desirable feature in my view.

**Comment by CombeeMike** [comment]
I would just like to add a tiny bit to what s3woz said in [above comment](https://github.com/nextcloud/desktop/issues/3244#issuecomment-1149717292):
> ... Anyway, syncing .lnk files with nextcloud is a very desirable feature in my view.

... which had been there for a very long time and around which users built many different workflows.

**Comment by ScarUY** [comment]
> that said, `.lnk` files should better stay out of a sync folder because there is no way to deal with them in the web UI or on other platforms (Androit mobile client, ...). They are not going to be really usable out of a desktop sync folder on Windows.

I hate to be "that guy" but this sounds a lot like some certain fruit-logo-that-doesn't-sell-fruits mentality where the company decides for the user and the user has no choice, with this attitude we'll end up syncing only pdf, doc, xls and a short etc...

**Comment by ChrisBiesterveld** [comment]
https://github.com/nextcloud/desktop/issues/3244#issuecomment-950726735

> Thanks all for your comments, showing the strong interest of the community to be able to sync lnk files again on Windows. @allexzander [mentioned that this needs QT6.2](https://github.com/nextcloud/desktop/pull/3057#issuecomment-893237979). This version has been [released end of September](https://wiki.qt.io/Qt_6.2_Release), but I guess Nextcloud needs time to work on its integration.
> 
> Waiting for that, @allexzander @FlexW may I ask you to reopen this ticket, as this issue is not solved yet, and I understand (with thanks ;) ), that it's not a "won't fix" issue?

It was already mentioned, the feature would return. Still not seeing it active, but I guess we just have to be patient ?

**Comment by chylex** [comment]
This is an issue that only affects virtual file systems, right? If waiting for a QT fix and then upgrading QT is taking over a year, and the problem is "Users who were not using VFS in the first place cannot sync .lnk files anymore", why can't the solution be allowing .lnk syncing as long as the user is not using VFS?

**Comment by pierre-pizzetta** [comment]
I also have users needing this feature, please bring .lnk files sync back. Like it said before, those shortcuts are part of a lot of people workflow for efficiency. Thanks !

**Comment by mgallien** [comment]
I understand that this issue is important for quite a few people
this is why I started working again on it

first the changes we would need in Qt will probably happen only for the next major version (they tend to not do changes breaking existing code within a major release, so all Qt6 versions will probably keep the same behavior they have in Qt5)

that said, I tested again with other cloud file sync and share tools
as far as I can say `.lnk` are not handled like normal files but rather being always there (they can never be freed entirely)

so I am under the impression that they might have the same issue we have

and we could most probably apply the same solution (so .lnk file could be either always offline or downloaded but never online only)

**Comment by biva** [comment]
Thank you very much @mgallien to spend time on this :) I'm not sure to fully understand you:
- Do we need to wait for changes within Qt to do something? If yes, how can we push Qt to do the required changes?
- What do you mean by _either always offline or downloaded but never online only_? 
  - I understand that we cannot apply "Free up local space" to this file, right?
  - What about if we "free up" the folder containing this file? The whole folder will be freed, except the lnk file which will remain locally, right? I think it's a very acceptable solution because these files are extremely light.
  
Does it require a lot of work to implement this solution?

Thank you again, sorry to not be able to help with the coding, but happy to test.

**Comment by julll190** [comment]
Hi,
I also need to sync several .lnk files, that is the main reason why I want to use Nextcloud on different computers, hope it will come back soon.
Thank you to accept to sync it again :)

**Comment by rexhunt** [comment]
Just dropping in to add my vote for this feature to return.
It sounds like it's more complicated than this, but an option to "treat .lnk files as regular files" would allow us to use Nextcloud within the Autodesk workflow that we've got.

Thank you for the work on this problem.

**Comment by n-aceves** [comment]
Hello, Please allow us to sync .lnk files again. Keeping it off by default but allowing us to have the option to enable the syncing of lnk files seems like the best solution.

Thanks.

**Comment by dinomueller** [comment]
I too need the syncing of .lnk files. I had several hundreds of them in my dropbox synced over several devices. Then I switched to Nextcloud and had to learn after client 3.2.1 it doesn't work anymore. So I had to go back to dropbox because of that, but I would like to return to Nextcloud again.

.lnk files are files! Not symlinks! I don't understand why you don't simply let us just sync them? They don't cause any problems on the server side.

**Comment by ChrisBiesterveld** [comment]
Thanks for bringing the functionality back !
Could you let us know in what build it will be provided ? I updated to Version 3.6.0 (Windows) today, but lnk files are still ignored.

**Comment by allexzander** [comment]
> Thanks for bringing the functionality back ! Could you let us know in what build it will be provided ? I updated to Version 3.6.0 (Windows) today, but lnk files are still ignored.

Will be released in 3.6.1

**Comment by mathew-cook** [comment]
Many thanks!

**Comment by Schniefel** [comment]
I am not entirely sure if this is fixed completely, as I have some .lnk-files which simply don't get synced with the desktop client. If I drag them in the web-ui the files get synced. So I assume there is a problem for the desktop client detecting some .lnk-files for syncing.

**Comment by Sieboldianus** [comment]
In the latest `3.15.0`, it seems `*.lnk` files are not synced anymore. I get errors for all my shortcuts.

**Comment by RokeJulianLockhart** [comment]
> https://github.com/nextcloud/desktop/issues/3244#issuecomment-2505374564

@Sieboldianus, got a log and/or screenshot of the errors?

**Comment by ChrisBiesterveld** [comment]
> In the latest `3.15.0`, it seems `*.lnk` files are not synced anymore. I get errors for all my shortcuts.

Fyi : Just did a test. I have 3.15 installed and everything seem to work fine.

**Comment by Sieboldianus** [comment]
I tested and I can create _new_ `.lnk` files. This morning, all errors appeared with existing `*.lnk` files that could not be synced from remote. I deleted these and the errors were gone. I will report if my newly created lnk files today will be successfully synced on me 2nd laptop tomorrow!

**Comment by RokeJulianLockhart** [comment]
> I tested and I can create new `.lnk` files. This morning, all errors appeared with existing `*.lnk` files that could not be synced from remote. I deleted these and the errors were gone.

@Sieboldianus, if you still retain those unsychronisable `.lnk` files in [`shell:RecycleBinFolder`](shell:RecycleBinFolder), try:

1. restoring one,
2. verifying that it's still unsychronisable,
3. creating a new one with the same parameters, and
4. verifying that the new one can be synchronized.

If it can be, then try `diff`-ing them to see what differs between them. If nothing, then it's probably a bug with [the "Files On-Demand" API](https://github.com/MicrosoftDocs/feedback/issues/1484#issuecomment-2506480742).

**Comment by Sieboldianus** [comment]
I just verified on one of my computers that

- I can restore one of the `*.lnk` files (a folder) that failed syncing yesterday; it restored, is accessible (works), and can delete it and create a new shortcut that syncs with the same parameters

For the file-shortcuts that failed on my other computer today, I will test tomorrow. But I have the feeling that it will be the same pattern.

**Comment by Sieboldianus** [comment]
Some more tests:
- I cannot create any lnk files (files or folder shortcuts) that do not produce these placeholder errors
- all lnk files are actually synced to the server, despite error

**Comment by dago666** [comment]
Same problem here.
*.lnk files are synchronized for the first time. But after client restart a sync error occur with 3.15.0.

After a downgrade to 3.14.0 it works again.
